File: cpp/tests/feed_sequence_recovery.cpp
#include "llab/feed_sequence_recovery.hpp"

#include <stdexcept>
#include <vector>

namespace {
void require(bool condition) {
    if (!condition)
        throw std::runtime_error("feed recovery test failed");
}
} // namespace

int main() {
    llab::FeedSequenceRecovery<int> recovery(4);
    std::vector<int> ready;

    recovery.begin_snapshot(10);
    require(recovery.on_incremental(12, 12, ready) == llab::FeedIncrementalResult::Buffered);
    require(recovery.on_incremental(11, 11, ready) == llab::FeedIncrementalResult::Buffered);
    require(recovery.on_incremental(11, 111, ready) == llab::FeedIncrementalResult::Buffered);
    require(recovery.finish_snapshot(ready));
    require((ready == std::vector<int>{11, 12}));
    require(recovery.expected_sequence() == 13);

    ready.clear();
    require(recovery.on_incremental(12, 12, ready) == llab::FeedIncrementalResult::IgnoreStale);
    require(recovery.on_incremental(14, 14, ready) == llab::FeedIncrementalResult::Gap);
    require(recovery.state() == llab::FeedRecoveryState::AwaitingSnapshot);

    recovery.begin_snapshot(20);
    require(recovery.on_incremental(22, 22, ready) == llab::FeedIncrementalResult::Buffered);
    require(!recovery.finish_snapshot(ready));
    require(recovery.state() == llab::FeedRecoveryState::AwaitingSnapshot);

    recovery.begin_snapshot(30);
    require(recovery.on_incremental(31, 31, ready) == llab::FeedIncrementalResult::Buffered);
    require(recovery.on_incremental(32, 32, ready) == llab::FeedIncrementalResult::Buffered);
    require(recovery.on_incremental(33, 33, ready) == llab::FeedIncrementalResult::Buffered);
    require(recovery.on_incremental(34, 34, ready) == llab::FeedIncrementalResult::Buffered);
    require(recovery.on_incremental(35, 35, ready) == llab::FeedIncrementalResult::BufferFull);
    require(recovery.state() == llab::FeedRecoveryState::AwaitingSnapshot);
}

File: rust/tests/feed_sequence_recovery.rs
use low_latency_lab_benchmarks::feed_sequence_recovery::{
    FeedSequenceRecovery, IncrementalResult, State,
};

#[test]
fn replays_a_contiguous_suffix_and_ignores_duplicates() {
    let mut recovery = FeedSequenceRecovery::<u64>::new(4);
    let mut ready = Vec::new();
    recovery.begin_snapshot(10);
    assert_eq!(recovery.on_incremental(12, 12, &mut ready), IncrementalResult::Buffered);
    assert_eq!(recovery.on_incremental(11, 11, &mut ready), IncrementalResult::Buffered);
    assert_eq!(recovery.on_incremental(11, 111, &mut ready), IncrementalResult::Buffered);
    assert!(recovery.finish_snapshot(&mut ready));
    assert_eq!(ready, vec![11, 12]);
    assert_eq!(recovery.expected_sequence(), Some(13));
    assert_eq!(recovery.on_incremental(12, 12, &mut ready), IncrementalResult::IgnoreStale);
    assert_eq!(recovery.on_incremental(14, 14, &mut ready), IncrementalResult::Gap);
    assert_eq!(recovery.state(), State::AwaitingSnapshot);
}

#[test]
fn refuses_partial_replay_and_buffer_overflow() {
    let mut recovery = FeedSequenceRecovery::<u64>::new(4);
    let mut ready = Vec::new();
    recovery.begin_snapshot(20);
    assert_eq!(recovery.on_incremental(22, 22, &mut ready), IncrementalResult::Buffered);
    assert!(!recovery.finish_snapshot(&mut ready));
    assert_eq!(recovery.state(), State::AwaitingSnapshot);

    recovery.begin_snapshot(30);
    for sequence in 31..=34 {
        assert_eq!(
            recovery.on_incremental(sequence, sequence, &mut ready),
            IncrementalResult::Buffered
        );
    }
    assert_eq!(recovery.on_incremental(35, 35, &mut ready), IncrementalResult::BufferFull);
    assert_eq!(recovery.state(), State::AwaitingSnapshot);
}
